proposals/xxxx-initializer-lists.md

```diff
@@ -0,0 +1,169 @@
+initializer has less arguments than the target has fields, the remaining fields
+are zero initialized. It is an error if an initializer list has too many
+arguments, if an initializer has arguments that cannot be converted to the
+destination type, or if an initializer list has incorrect nested braces.
+
+
+The following examples which are valid in C++ are not valid in HLSL:
+
+```c++
+struct A {
+  int a;
+  double b;
+};
+struct B {
+  A a[2];
+  int c;
+};
+B b = {};
+A a = {0};
+B b2 = {{{0}}, 0};
+B b3 = {{0}, 0};
+```
+
+Given the differences in the language described above it is likely too
+significant of a change to make Clang follow C/C++ behavior without a transition
+period. Since Clang and DXC will have different underlying representations for
+resources and other built-in types, it is likely extremely difficult to make DXC
+and Clang match on a new behavior. It will also be difficult to provide
+migration tooling in DXC. For these reasons it is proposed to solve this problem
+in Clang.
+
+## Proposed solution
+
+This solution is based on an assumption that we will adopt
+[Proposal 0005 - Strict Initializer Lists](https://github.com/microsoft/hlsl-specs/blob/main/proposals/0005-strict-initializer-lists.md)
+for HLSL 202y. The strict initializer lists proposal is motivated by a closer
+alignment with C++, better error handling, and resolving incompatibilities with
+variadic template initializer lists.
+
+To facilitate implementing C++ initializer list behavior while also providing
+compatibility with existing HLSL, this proposal suggests that Clang parse HLSL
+initialization syntax into valid C++ initializer list ASTs.
+
+This would mean that given an example like:
+
+```hlsl
+struct A {
+  int a;
+  double b;
+};
+struct B {
+  A a[2];
+  int c;
+};
+B b3 = {{1, {2, 3}}, {4, 5}};    // Completely random grouping of arguments
+```
+
+The initializer list for `b3` would be parsed into an AST something like:
+
+```
+InitListExpr
+  InitListExpr
+    InitListExpr
+      IntegerConstant 1
+      ImplicitCastExpr int->double
+        IntegerConstant 2
+    InitListExpr
+      IntegerConstant 3
+        ImplicitCastExpr int->double
+          IntegerConstant 4
+  IntegerConstant - 5
+```
+
+This AST structure allows the HLSL initializers to be implemented without any
+need to change code generation. In an alternate example the AST can contain
+member accesses with type checking of each conversion as well. Take the
+following code and proposed AST:
+
+```hlsl
+float4 f = 1.0.xxxx;
+struct Blah {
+  int a, b;
+  half c , d;
+};
+
+Blah blah = {f};
+```
+
+```
+InitListExpr
+  ImplicitCastExpr float->int
+    ExtVectorElementExpr .x
+      DeclRefExpr F
+  ImplicitCastExpr float->int
+    ExtVectorElementExpr .y
+      DeclRefExpr F
+  ImplicitCastExpr float->half
+    ExtVectorElementExpr .z
+      DeclRefExpr F
+  ImplicitCastExpr float->half
+    ExtVectorElementExpr .w
+      DeclRefExpr F
+```
```

Comment:
Rename `F` as `f` following changes in commit 690234038d470a147ec4205f027fc367cc042d25